File: Leetcode/0001-1000/0601-0700/0624. Maximum Distance in Arrays - One Scan.py
# https://leetcode.com/problems/maximum-distance-in-arrays/

from typing import List


class Solution:
    def maxDistance(self, arrays: List[List[int]]) -> int:
        curMax, curMin, result = arrays[0][-1], arrays[0][0], 0
        for i in range(1, len(arrays)):
            result = max(result, abs(arrays[i][0] - curMax), abs(arrays[i][-1] - curMin))
            curMax = max(curMax, arrays[i][-1])
            curMin = min(curMin, arrays[i][0])
        return result

File: Leetcode/0001-1000/0601-0700/0624. Maximum Distance in Arrays - One Scan.cpp
// https://leetcode.com/problems/maximum-distance-in-arrays/

static int fast_io = []() {
    std::ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    return 0;
}();

class Solution {
public:
    int maxDistance(vector<vector<int>>& arrays) {
        int curMax = arrays[0].back();
        int curMin = arrays[0].front();
        int result = 0;
        for (int i = 1; i < arrays.size(); i++) {
            result = max({result, abs(arrays[i].front() - curMax), abs(arrays[i].back() - curMin)});
            curMax = max(curMax, arrays[i].back());
            curMin = min(curMin, arrays[i].front());
        }
        return result;
    }
};